### ISSUE TITLE
esp8622 crash..

### ISSUE THREAD

**Issue by wchpikus** [nailbuster/esp8266FTPServer#24]
esp with working wifi, www server, spiffs, socket, oled via spi.
added ftp, and when transfering file esp crash.
using filezilla, disconnect client when choiced transfer file (i have to connect again, wait for connection).

**Comment by Monarch73** [comment]
@wchpikus 
do you mind trying out my ESP-FTP-Server? Its a single .h-File that you can copy and paste into your project. It just depends on the AsyncTCP-Library from @me-no-dev

Please note that it is required that you limit the number of concurrent connections to 1 in filezilla. Filetransfers are done via PORT-Command, so it supports non-passive mode only.

https://github.com/Monarch73/MicroHomebridgeAlexaEsp8266/blob/master/MicroHomebridgeAlexaEsp8266/FtpServ.h

**Comment by wchpikus** [comment]
hi

ESP8266FtpServer.h and ESP8266FtpServer.cpp files, and setup in main and setup in ino.
connections is only1 in filezilla and non-passive mode.

I can remove, see dir and rename, but upload made crash:(
you write about this file FtpServ.h?
 i can try it.

**Comment by Monarch73** [comment]
@wchpikus 
You seem to misunderstand... I am not the author of THIS (esp8266FTPServer) project. But I had trouble using THIS library as well and ended up writing my own ftp-server

**Comment by wchpikus** [comment]
yes, i corrected post:)

now i will thy our ftp.

Could you send me some more info about that?

e.g how to use?


FtpServ(char *username, char *password) = FtpServ(test,test); i see, but do you need inicialize or handle something?

**Comment by Monarch73** [comment]
```
FtpServ *srv = new FtpServ("username","password");
srv->begin();
```
Put that code somewhere in the setup()
Note that for now username and password is ignorened when you login. Security was not my concern, when I created this...it'll simply accept any user/pass you give it.

**Comment by wchpikus** [comment]
https://github.com/me-no-dev/ESPAsyncTCP/tree/master/src
can i get it ESPAsyncTCP.h and ESPAsyncTCP.cpp from this link?

**Comment by Monarch73** [comment]
https://github.com/me-no-dev/ESPAsyncTCP/archive/master.zip

download this zip and in arduino studio you click "sketch" -> "library" -> add from zzip

**Comment by wchpikus** [comment]
ok i now test it..
is it works with spiffs?

**Comment by Monarch73** [comment]
yes, it works with spiffs. Please note though, that it only has very little error handling. It assumes, that spiffs is formatted and ready to go. It won't work if spiffs is not available.

**Comment by wchpikus** [comment]
ok work, but not all


data received from client 192.168.11.2 
PWD

 new client has been connected to server, ip: 192.168.11.2
 data received from client 192.168.11.2 
USER volvoaux
sending user

 data received from client 192.168.11.2 
PASS wchpikus
sending pass

 data received from client 192.168.11.2 
CWD /

 data received from client 192.168.11.2 
TYPE I

 data received from client 192.168.11.2 
PASV

ok i try send some file


 data received from client 192.168.11.2 
SIZE c8_d7_b0_1a_34_e4.txt

 data received from client 192.168.11.2 
MDTM c8_d7_b0_1a_34_e4.txt

 data received from client 192.168.11.2 
TYPE A

 data received from client 192.168.11.2 
PASV

 data received from client 192.168.11.2 
TYPE I

 data received from client 192.168.11.2 
PASV

 data received from client 192.168.11.2 
SIZE c8_d7_b0_1a_34_e4.txt

 data received from client 192.168.11.2 
MDTM c8_d7_b0_1a_34_e4.txt

 data received from client 192.168.11.2 
TYPE A

 data received from client 192.168.11.2 
PASV

 data received from client 192.168.11.2 
TYPE I

 data received from client 192.168.11.2 
PASV

 data received from client 192.168.11.2 
SIZE c8_d7_b0_1a_34_e4.txt

 data received from client 192.168.11.2 
MDTM c8_d7_b0_1a_34_e4.txt

 data received from client 192.168.11.2 
TYPE A

 data received from client 192.168.11.2 
PASV

 data received from client 192.168.11.2 
TYPE I

 data received from client 192.168.11.2 
PASV

 data received from client 192.168.11.2 
PWD


from filezila:

Status:	Starting upload of G:\oled animation\OLED-esp-aux_phone_english_www_spiffs_ftp_v6\c8_d7_b0_1a_34_e4.txt
Status:	Retrieving directory listing of "/"...
Command:	TYPE I
Response:	200 JAAAA MANNN!
Command:	PASV
Response:	500 Unknown command.
Command:	SIZE c8_d7_b0_1a_34_e4.txt
Response:	500 Unknown command.
Command:	MDTM c8_d7_b0_1a_34_e4.txt
Response:	500 Unknown command.
Command:	TYPE A
Response:	200 JAAAA MANNN!
Command:	PASV
Response:	500 Unknown command.
Error:	File transfer failed
Status:	Retrieving directory listing of "/"...
Command:	TYPE I
Response:	200 JAAAA MANNN!
Command:	PASV
Response:	500 Unknown command.
Error:	Failed to retrieve directory listing
Status:	Sending keep-alive command
Status:	Sending keep-alive command
Status:	Sending keep-alive command

**Comment by wchpikus** [comment]
Spiifs ready, 4M ( 3M spiffs for data).

**Comment by Monarch73** [comment]
Hey, at least you see no crashes :-). But I just realized forgot to meantion a few more things:

- PASV is not supported. Use PORT instead (non-passive-mode only). Use the connection manager in Filezilla to configure this. There you also limit the number of concurrent connections to 1
- Subdirectories in spiffs are not supported.. You can only up/download to main directory on ESP
- Length of filenames is limited to 32 characters in spiffs...

**Comment by Monarch73** [comment]
also please make sure, that you don't use any connection security features..
SIZE and MDTM are not supported...(and not required)

**Comment by wchpikus** [comment]
**ok setuped to activ:**

Status:	Connecting to 192.168.11.1:21...
Status:	Connection established, waiting for welcome message...
Status:	Server does not support non-ASCII characters.
Status:	Logged in
Status:	Retrieving directory listing...
Status:	Directory listing of "/" successful
Status:	Disconnected from server
Status:	Connecting to 192.168.11.1:21...
Status:	Connection established, waiting for welcome message...
Status:	Server does not support non-ASCII characters.
Status:	Logged in
Status:	Starting upload of G:\oled animation\OLED-esp-aux_phone_english_www_spiffs_ftp_v6\c8_d7_b0_1a_34_e4.txt
Status:	File transfer successful, transferred 17 140 bytes in 1 second
Status:	Retrieving directory listing of "/"...
Status:	Directory listing of "/" successfu


**ok but when i send 160kb file...crash:)**

dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

 dataconnection received data from 192.168.11.2 

Exception (0):
epc1=0x4020ab3c epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000

ctx: sys 
sp: 3ffff790 end: 3fffffb0 offset: 01a0

>>>stack>>>
3ffff930:  4000050c 07672d1f 401021f2 00000100  
3ffff940:  400043df 00000030 3fffc248 4000050c  
3ffff950:  40004225 00000030 00000014 ffffffff  
3ffff960:  4000420c 00000000 3ffffa40 08000000  
3ffff970:  60000200 00000002 02000000 02000000  
3ffff980:  02000000 00000018 04000102 60000200  
3ffff990:  3ffffb80 00000010 001543c8 00000030  
3ffff9a0:  00000000 4000422e 60000200 3ffffa24  
3ffff9b0:  00000100 40004aa0 000000fc 00153d04  
3ffff9c0:  00000100 3ffffa20 3fffc718 00153d04  
3ffff9d0:  00000004 4010551e 40105a64 00000004  
3ffff9e0:  3ffffa20 3ffffa20 00000000 3ffffa64  
3ffff9f0:  3fffc718 3ffffa20 000000fc 00154104  
3ffffa00:  402170e1 402170d6 00153d04 00000004  
3ffffa10:  00153d05 00153d04 00000001 40219f20  
3ffffa20:  ffffff7e 3ffffa60 00000000 000000c4  
3ffffa30:  00000000 4000444e 0000000c c0000000  
3ffffa40:  00000002 4000420c 60000200 3ffffb80  
3ffffa50:  00000100 40004aa0 000000f8 00154308  
3ffffa60:  00000100 3ffffac0 3fffc718 00154308  
3ffffa70:  000000d0 4010551e 40105a64 000000d0  
3ffffa80:  3ffffac0 3ffffac0 00154308 000000e7  
3ffffa90:  3fffc718 3ffffac0 60000200 3fff5bd4  
3ffffaa0:  00154308 402170d6 00154308 000000d0  
3ffffab0:  001543d8 001543d8 000000d3 40219ed8  
3ffffac0:  41746857 7858464b 6e684943 68200a0d  
3ffffad0:  74343449 32566c57 43437175 58326543  
3ffffae0:  5437694f 446e4e56 53563541 7678774d  
3ffffaf0:  4a636f6e 39335855 616e7870 314c746f  
3ffffb00:  58623258 7842772f 45356962 6d527468  
3ffffb10:  4535764a 79494568 5a200a0d 776f7049  
3ffffb20:  597a5164 34416937 65374653 75655a56  
3ffffb30:  36776552 442b6d6e 38684a38 66786358  
3ffffb40:  514c7571 6266766f 76686144 5243624a  
3ffffb50:  74494945 4d453870 64786a73 544c3650  
3ffffb60:  7454634f 37200a0d 6b613651 666d7865  
3ffffb70:  42384347 494a4b6c 45655438 61523162  
3ffffb80:  6c543852 4b394750 774c6330 4d5a4979  
3ffffb90:  41702f58 74515671 65326144 73516855  
3ffffba0:  52675356 6c657577 4c2f4d70 4770616b  
3ffffbb0:  4b323731 53625246 00000000 00000029  
3ffffbc0:  3fff417c 00154000 3fff32f0 3fff4190  
3ffffbd0:  3fff417c 00154000 3fff32f0 40206d95  
3ffffbe0:  00000100 3ffffcd3 00000000 3fff41a4  
3ffffbf0:  00000000 40101f8c 3fffc200 4021f388  
3ffffc00:  3ffffd30 3ffe8304 3ffffc90 4021b2f1  
3ffffc10:  3ffeacf5 3fff496c 3ffffc40 40204710  
3ffffc20:  40101be1 3ffee6e0 00000000 3ffffc90  
3ffffc30:  4021f388 3ffffd30 3fff496c 4021b6ac  
3ffffc40:  00000100 3fff3aac 00000000 00000001  
3ffffc50:  3ffe8304 0767c0c1 401021f2 00000100  
3ffffc60:  00000000 400042db 3ffeb570 00000100  
3ffffc70:  3fff32f0 00054000 0000002a 40204710  
3ffffc80:  00000000 00000080 00000000 00002e23  
3ffffc90:  00000001 00002e22 00000000 0000ffff  
3ffffca0:  3fff3aac 00000001 00000020 00000100  
3ffffcb0:  00000002 00000000 00000002 40219dbc  
3ffffcc0:  3fff5ccf 000000d0 000000d0 00000200  
3ffffcd0:  3fff5cd2 3ffffd30 3fff32f0 40204b31  
3ffffce0:  00000000 00000000 3ffffd34 000000d3  
3ffffcf0:  3fff5ccf 00154305 3fff32f0 40206e94  
3ffffd00:  3fff4190 00000007 3fff417c 3fff42b8  
3ffffd10:  3fff5b8a ffffffff ffffffff 3fff32f0  
3ffffd20:  00000300 00000100 3ffffd90 40204d98  
3ffffd30:  00000002 0005002a 3fff32f0 0000004a  
3ffffd40:  3fff0018 3fff5ccf 000000d3 00000000  
3ffffd50:  00000001 00002000 00100000 402212e8  
3ffffd60:  00000008 3fff2ab8 00000000 3fff32f0  
3ffffd70:  00000003 3fff39ac 3fff40b4 4020593c  
3ffffd80:  00000001 3ffffd98 3fff2bf0 40227100  
3ffffd90:  01770018 0521eefc 05410542 00000521  
3ffffda0:  00000145 00016e68 00000177 00000000  
3ffffdb0:  00000003 00000002 00000000 000000d3  
3ffffdc0:  00000218 3fff5b8a 00000175 00000100  
3ffffdd0:  3fff16e8 00000008 3ffffde0 3fff5b8a <
3ffffde0:  00016e68 00000218 00016e68 402031e5  
3ffffdf0:  63657220 65766965 61642064 66206174  
3ffffe00:  206d6f72 2e323931 2e383631 3fff5b8a  
3ffffe10:  00000218 00016e68 3fff32f0 402038fc  
3ffffe20:  3fff40b4 3ffffe30 00000008 402184b5  
3ffffe30:  3fff40b4 00000218 3fff496c 0000ff00  
3ffffe40:  3fff16c0 0000038f 0000038f 00000000  
3ffffe50:  3fff5894 00000218 3fff4f2c 402194c5  
3ffffe60:  2e323931 2e383631 3fff5b8a 4021a4d1  
3ffffe70:  3fff5894 00000218 3fff5b8a 4020ab14  
3ffffe80:  00000000 00000000 00000000 401071f0  
3ffffe90:  020ba8c0 00000008 3ffef928 00000030  
3ffffea0:  3fff5894 3fff5b44 00000000 4021a096  
3ffffeb0:  3fff0000 00000000 3fff2a4c 40215e50  
3ffffec0:  3fff5974 00000001 3ffeab24 00000064  
3ffffed0:  00000860 3ffea988 3ffea988 3fff2a6c  
3ffffee0:  3fff5974 3fff2a50 3fff2a4c 40215e94  
3ffffef0:  3fff5974 3fff2a50 3fff2a51 40223cd4  
3fffff00:  010ba8c0 0000000a 3ffea988 00000000  
3fffff10:  00000000 0000024e 3fffff80 3fff5b62  
3fffff20:  3fff2bf0 3fff5b44 3fff2cd8 40227c01  
3fffff30:  00000014 0000038f 0000038f 3fff2bf0  
3fffff40:  3fffdc80 3fff3194 3fff496c 3fff3214  
3fffff50:  00000008 3fff2bf0 3fff5b44 40221259  
3fffff60:  3fffdc80 3fff3194 3fff496c 4010453c  
3fffff70:  4023e7fe 3fff3194 3fff496c 4023e810  
3fffff80:  3fff5b54 3fff5b44 00000000 3fffdcb0  
3fffff90:  3ffeffd8 00000000 3fff496c 40240773  
3fffffa0:  40000f49 3fffdab0 3fffdab0 40000f49  
<<<stack<<<

**btw, Status:	Retrieving directory listing of "/"... listing works**
**Small file work well:**

File system (SPIFFS): 
    Total KB: 2880.13 KB
    Used KB: 19.85 KB
    Block size: 8192
    Page size: 256
    Maximum open files: 5
    Maximum path length: 32

SPIFFS directory {/} :
**/c8_d7_b0_1a_34_e4.txt / 17140**

**Comment by Monarch73** [comment]
I don't think that crash is related to the ftp-server code. It must be something in the sdk. Do you know what version of sdk you use?

**Comment by wchpikus** [comment]
How to check it in arduino?

**Comment by wchpikus** [comment]
AppData\Local\Arduino15\packages\esp8266\hardware\esp8266\2.4.1

**Comment by Monarch73** [comment]
2.4.1 That should do....even thou you should be able to update to 2.4.2...but I don't think that will fix the problem. Let me check when I am back at my desk...I must admit that I never tested this with files >100k

**Comment by Monarch73** [comment]
In the meantime, you can try to reformat spiffs...might be corrupted or something...

**Comment by wchpikus** [comment]
I formated few times..it is no flash problem;)

Please try it and write..and what sdk have you.

**Comment by Monarch73** [comment]
I always work with the latest release, so its 2.4.2 for me.... But I also work with Visual Studio Community with VMicro plugin, so I would not know how you can update your Arduino Studio....

**Comment by Monarch73** [comment]
another thing you can try: Create a new project with just the ftpServer-code in it. This way you make sure that no other parts of your code are causing the crash...

**Comment by wchpikus** [comment]
Ok i try.

**Comment by wchpikus** [comment]
give me some time, i have to update and install new release:)

**Comment by wchpikus** [comment]
ng core 'esp8266' from platform in folder: AppData\Local\Arduino15\packages\esp8266\hardware\esp8266\2.4.2

ok updated, i test it now...

**Comment by Monarch73** [comment]
I have bad news for you. I just uploaded a 300KB File to my nodemcu-Module with no problems...

I used the following code

```
FtpServ *ftpServer;
void setup() {
	Serial.begin(115200);
	delay(20);
	SPIFFS.begin();
	// create access point
	WiFi.mode(WIFI_STA);
	WiFi.begin("xxxx", "xxxx");

	while (WiFi.status() != WL_CONNECTED) {
		delay(500);
		Serial.print(".");
	}

	Serial.println("IP address: ");
	Serial.println(WiFi.localIP());
	ftpServer = new FtpServ("1", "2");
	ftpServer->begin();
}

// the loop function runs over and over again until power down or reset
void loop() {
  
}
```

**Comment by wchpikus** [comment]
now i disabling all process (www, mdns..)to eliminate problem..
please wait:)

**Comment by Monarch73** [comment]
I just noticed that ftp-code does not do a SPIFFS.begin()....that needs to be done in the setup()-function as well

**Comment by wchpikus** [comment]
SPIFFS.begin() I have..
This is not a problem..some of my function destruct stack..
i have find what..
now i'll test your code.

**Comment by wchpikus** [comment]
hmm,  you have right...your code works well, so i have to find what do not want to work with this..
i thing this is connected with #<WiFiClient.h> and client.connect(gateway,httpPort) (i listening on 23 port from other device)..
i try to disable it when using ftp..

**Comment by wchpikus** [comment]
Did you test of size of file after and before upload?
I have some strange value

**Comment by Monarch73** [comment]
Nope...the code does not check filesizes...its working fine for me, too...

also, having multiple connections should generally not be an issue.  Maybe you want to try a different board?

**Comment by wchpikus** [comment]
Only your code works on the same board, so this is isue of rest of my firmware..but i will check it all, disable one by one..

**Comment by Monarch73** [comment]
ok..._maybe_ you want to consider changeing your code to use AsyncTCP too.

a bit of technical background:
Synchronous I/O in a realtime environment like the ESP8266 can be problematic. When execution-path does not leave the loop()-function from time to time, the Realtime-Operating-system of the board will not have enough time to take care of the TCP-Stack.

This is why my loop()-function has as little to do as possible, and I keep recommending this to people...do not block the operating-system from getting its job done!

**Comment by wchpikus** [comment]
i see, but you know, i have oled spi,telnet (two ways),and it could be no time, or interrupt some function, when receive data and got some data from other socket (telnet).
i check it, i have lot of work with them, but i see your code works..
Maybe this first also works, it must be clear, for future:)

**Comment by wchpikus** [comment]
did you see difference between generic and nodemcu?

**Comment by Monarch73** [comment]
what you mean "generic"? What board do you have? (amazon-link)

**Comment by wchpikus** [comment]
nodemcu (aliexpress)
i write about compile to generic esp module vc nodemcu.

**Comment by wchpikus** [comment]
i disable all function from loop, still crash.
could you (if you have experience, in setup i connect client (client.connect(gateway,httpPort) ) to 23 port.
and do not disconnect them..
your ftp make new connection to port 21 to client or different port?
i thing this is problem, nothing more i do not have run when client connecting to ftp server.

**Comment by Monarch73** [comment]
where is "FtpServ * srv=new FtpServ(...)"-line  in your code? Inside  setup() or outside? Try to move the declaration in a global scope-right below the #include-lines.....

**Comment by Monarch73** [comment]
I can't believe that the port-number could cause these kind of problems....

**Comment by wchpikus** [comment]
where is "FtpServ * srv=new FtpServ(...) is de fact in setup, but on stand alone function.

**Comment by wchpikus** [comment]
but it must be below spiffs.begin() so this is not good to move there?

**Comment by Monarch73** [comment]
follow the way I did it....

FtpServ *ftpServer;  <--- this goes below the include lines

ftpServer=new FtpServ(....
ftpServer->begin()           <---- these two lines in the setup

**Comment by Monarch73** [comment]
I am suspecting that the ftpServer instance gets picked up by the garbage collector....

**Comment by wchpikus** [comment]
i test it now..sorry, but after upgrade sdk compilation was terrible slow:(

**Comment by Monarch73** [comment]
I could guide you through the installation of visual studio community 2017 and Visual Micro Plugin. Its a really comfortable development environment and its free

**Comment by wchpikus** [comment]
yes...i hate arduino..
Eclipse of visual studio..or some normal editor:)

**Comment by Monarch73** [comment]
https://visualstudio.microsoft.com/de/thank-you-downloading-visual-studio/?sku=Community&rel=15
 and
https://www.python.org/downloads/release/python-2715/

these are the two things, that you need to install....

**Comment by Monarch73** [comment]
make sure, you tell visual studio setup to install the c++-development component

**Comment by wchpikus** [comment]
ok, and toolchain for esp?
can i install from zip file or?

**Comment by Monarch73** [comment]
in the workload menu of the setup, just select  
![image](https://user-images.githubusercontent.com/2189402/48921810-8e0b0680-eea2-11e8-90af-ef2960ceb054.png)

**Comment by wchpikus** [comment]
https://www.visualmicro.com/page/esp8266.aspx
ok tnx, and this for esp

**Comment by wchpikus** [comment]
Could i import project from arduino to visual?
Offcource ino to c rename..

**Comment by Monarch73** [comment]
to install the arduino sdk for ESP later via git:

- create some empty directory
- create a directory called "hardware"
- inside hardware, you create another directory "esp8266"
- inside "esp8266" directory you execute git-command (git bash) : git clone --recursive https://github.com/esp8266/Arduino.git esp8266 
- switch to release code by cd-ing into esp8266 and run: git checkout tags/2.4.2
- In it, you'll find a directory called tools...
- execute "python get.py" in command line

**Comment by Monarch73** [comment]
unfortunatly, there is no easy way to import arduino stuido project into visual studio....you will have to create a new visual-micro project in Visual studio and copy and paste the files manually

**Comment by wchpikus** [comment]
Sure, it is no problem to copy;)
I sow that is possibility to debug code,is it true?
Could we talk to priv, my @ wchpikus@gmail.com.
Big tnx man for help, is it hard to find someone who can and like to help with esp;)

**Comment by Monarch73** [comment]
I just sent you a mail...

**Comment by wchpikus** [comment]
to gmail? try to wchpikus@wp.pl also

**Comment by wchpikus** [comment]
i tried on ccp do this:
U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI *u8g2;

void init_screen(){
  u8g2 = new U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 15, /* dc=*/ 2, /* reset=*/ 5);
  u8g2->begin(); // use -> instead of . to access methods now
}
and the h file:

extern U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI *u8g2;
void init_screen();

and:
init_screen();  in main file..
and first function u8g2.setContrast(backligth); compile error..

request for member 'setContrast' in 'u8g2', which is of pointer type 'U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI*' (maybe you meant to use '->' ?) 

Could you tell me hot to declare this ssd on other files in project?

**Comment by Monarch73** [comment]
`u8g2 = new U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=/ 15, / dc=/ 2, / reset=*/ 5);`
I assume you meant:
 
`u8g2 = new U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI(U8G2_R0, /* cs=/ 15, / dc=/ 2, / reset=*/ 5);`

**Comment by wchpikus** [comment]
U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 15, /* dc=*/ 2, /* reset=*/ 5); => ok

U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI(U8G2_R0, /* cs=/ 15, / dc=/ 2, / reset=*/ 5); =>'u8g2' was not declared in this scope after compilation for first function (u8g2.setDisplayRotation(U8G2_R2);)
